src/template.rs

```diff
@@ -0,0 +1,122 @@
+use std::fs::File;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use xml::attribute::OwnedAttribute;
+use xml::reader::XmlEvent;
+use xml::EventReader;
+
+use crate::error::TiledError;
+use crate::{util::*, EmbeddedParseResultType, MapTilesetGid, ObjectData, ResourceCache, Tileset};
+
+/// A template, consisting of an object and a tileset
+///
+/// The path is the path to the template file; it is used to identify them uniquely, and skip parsing the file multiple
+/// times.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct Template {
+    pub path: PathBuf,
+    pub tileset: Option<Arc<Tileset>>,
+    pub object: Option<ObjectData>,
+}
+
+impl Template {
+    pub(crate) fn parse_and_append_template(
+        path: &Path,
+        templates: &mut Vec<Template>,
+        cache: &mut impl ResourceCache,
+    ) -> Result<usize, TiledError> {
+        // First, check if this template already exists in our list of templates (it will be uniquely identified by its
+        // path).
+        if let Some((id, _)) = templates
+            .iter()
+            .enumerate()
+            .find(|(_, template)| template.path == path)
+        {
+            return Ok(id);
+        }
```

Comment:
This can be moved to the caller using `cache.get_or_try_insert_template` (or `get_template` and then `insert_template`)